pkg/image/registry/imagestream/etcd/image.go

```diff
@@ -0,0 +1,186 @@
+type ImageLayerIndex interface {
+	cache.SharedIndexInformer
+}
+
+type ImageStore interface {
+	rest.Watcher
+	rest.Lister
+}
+
+// NewImageLayerIndex creates a new index over a store that must return
+// images.
+func NewImageLayerIndex(store ImageStore) ImageLayerIndex {
+	ctx := apirequest.NewContext()
+	informer := cache.NewSharedIndexInformer(&cache.ListWatch{
+		ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+			obj, err := store.List(ctx, &metainternalversion.ListOptions{
+				ResourceVersion: options.ResourceVersion,
+				Limit:           options.Limit,
+				Continue:        options.Continue,
+			})
+			if err != nil {
+				return nil, err
+			}
+			list, ok := obj.(*imageapi.ImageList)
+			if !ok {
+				return nil, fmt.Errorf("unexpected store type %T for layer index", obj)
+			}
+			// reduce the full image list to a smaller subset.
+			out := &metainternalversion.List{
+				Items: make([]runtime.Object, len(list.Items)),
+			}
+			out.Continue = list.Continue
+			out.ResourceVersion = list.ResourceVersion
+			for i, image := range list.Items {
+				out.Items[i] = &ImageLayers{
+					Name:     image.Name,
+					Layers:   image.DockerImageLayers,
+					Manifest: manifestFromImage(&image),
+				}
+			}
+			return out, nil
+		},
+		WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+			w, err := store.Watch(ctx, &metainternalversion.ListOptions{
+				ResourceVersion: options.ResourceVersion,
+			})
+			if err != nil {
+				return nil, err
+			}
+			return watch.Filter(w, func(in watch.Event) (out watch.Event, keep bool) {
+				if in.Object == nil {
+					return in, true
+				}
+				// reduce each object to the minimal subset we need for the cache
+				image, ok := in.Object.(*imageapi.Image)
+				if !ok {
+					return in, true
+				}
+				in.Object = &ImageLayers{
+					Name:            image.Name,
+					ResourceVersion: image.ResourceVersion,
+					Layers:          image.DockerImageLayers,
+					Manifest:        manifestFromImage(image),
+				}
+				return in, true
+			}), nil
+		},
+	}, &ImageLayers{}, 0, cache.Indexers{
+		// layers allows fast access to the images with a given layer
+		"layers": func(obj interface{}) ([]string, error) {
+			entry, ok := obj.(*ImageLayers)
+			if !ok {
+				return nil, fmt.Errorf("unexpected cache object %T", obj)
+			}
+			keys := make([]string, 0, len(entry.Layers))
+			for _, layer := range entry.Layers {
+				keys = append(keys, layer.Name)
+			}
+			return keys, nil
+		},
+	})
+	return informer
+}
+
+// manifestFromImage attempts to find a manifest blob description from
+// an image. Images older than schema2 in Docker do not have a manifest blob.
+func manifestFromImage(image *imageapi.Image) *imageapi.ImageLayer {
+	if image.DockerImageManifestMediaType != "application/vnd.docker.distribution.manifest.v2+json" {
+		return nil
+	}
+	return &imageapi.ImageLayer{
+		Name:      image.DockerImageMetadata.ID,
+		MediaType: image.DockerImageManifestMediaType,
```

Comment:
The function name is misleading. And the return value too.

For v2 manifests `DockerImageMetadata.ID` contains a digest of [the image configuration](https://github.com/opencontainers/image-spec/blob/master/config.md). It differs from a manifest.

The MediaType for this blob is a though question, but it definitely shouldn't be DockerImageManifestMediaType.

Reply:
As LayerSize is not filled here, it shouldn't be used anywhere.

Reply:
I’ll probably say we just mark layer size as optional in the api.  It’s not fundamental for this call (although I would prefer it be correct). Is there a reason we didn’t record it on the image object?  Or did we just lose that info when we moved back to storing this in the registry?  We have it when we post a new image to the registry, correct?

You’re right that this should be the configuration mimetype.

Reply:
Probably can make LayerSize a pointer and have it be nil if unknown.

Reply:
Updated to make layer size be a pointer and nil if unknown.